File: rust-api/src/models/user_model.rs
use chrono::{DateTime, Utc};
use mongodb::bson::{oid, spec::BinarySubtype, Binary};
use serde::{Deserialize, Serialize};

use crate::services::password_service::hash_password;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct User {
    #[serde(rename = "_id")]
    pub id: oid::ObjectId,
    pub name: String,
    pub email: String,
    pub password: Binary,
    pub schema_version: u8,
    pub create_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PublicUser {
    pub id: String,
    pub name: String,
    pub email: String,
    pub create_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct PublicUserWithToken {
    pub id: String,
    pub name: String,
    pub email: String,
    pub create_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub token: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct CreateUser {
    pub name: String,
    pub email: String,
    pub password: String,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct LoginUser {
    pub email: String,
    pub password: String,
}

impl From<User> for PublicUser {
    fn from(user: User) -> Self {
        PublicUser {
            id: user.id.to_hex(),
            name: user.name,
            email: user.email,
            create_at: user.create_at,
            updated_at: user.updated_at,
        }
    }
}

impl User {
    pub fn new(user: CreateUser) -> Self {
        let hash = hash_password(&user.password);
        User {
            id: oid::ObjectId::new(),
            name: user.name,
            email: user.email,
            password: Binary {
                subtype: BinarySubtype::Generic,
                bytes: hash,
            },
            schema_version: 1,
            create_at: Utc::now(),
            updated_at: Utc::now(),
        }
    }
}

impl PublicUserWithToken {
    pub fn new(user: User, token: String) -> Self {
        PublicUserWithToken {
            id: user.id.to_hex(),
            name: user.name,
            email: user.email,
            create_at: user.create_at,
            updated_at: user.updated_at,
            token,
        }
    }
}

File: go-api/models/user_model.go
package models

import (
	"go-api/pkg"
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type User struct {
	ID           primitive.ObjectID `json:"_id" bson:"_id"`
	Name         string             `json:"name" bson:"name"`
	Email        string             `json:"email" bson:"email"`
	Password     []byte             `json:"password" bson:"password"`
	SchemaVerion uint8              `json:"schema_version" bson:"schema_version"`
	CreateAt     time.Time          `json:"create_at" bson:"create_at"`
	UpdatedAt    time.Time          `json:"updated_at" bson:"updated_at"`
}

type PublicUser struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Email     string `json:"email"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
}

type PublicUserWithToken struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Email     string `json:"email"`
	CreatedAt string `json:"created_at"`
	UpdatedAt string `json:"updated_at"`
	Token     string `json:"token"`
}

type CreateUser struct {
	Name     string `json:"name" binding:"required"`
	Email    string `json:"email" binding:"required,email"`
	Password string `json:"password" binding:"required"`
}

type LoginUser struct {
	Email    string `json:"email" binding:"required,email"`
	Password string `json:"password" binding:"required"`
}

func (u *User) PublicUser() PublicUser {
	return PublicUser{
		ID:        u.ID.Hex(),
		Name:      u.Name,
		Email:     u.Email,
		CreatedAt: u.CreateAt.String(),
		UpdatedAt: u.UpdatedAt.String(),
	}
}

func (u *User) PublicUserWithToken(token string) PublicUserWithToken {
	return PublicUserWithToken{
		ID:        u.ID.Hex(),
		Name:      u.Name,
		Email:     u.Email,
		CreatedAt: u.CreateAt.String(),
		UpdatedAt: u.UpdatedAt.String(),
		Token:     token,
	}
}

func (u *CreateUser) ToUser() User {
	return User{
		ID:           primitive.NewObjectID(),
		Name:         u.Name,
		Email:        u.Email,
		Password:     pkg.HashPassword(u.Password),
		SchemaVerion: 1,
		CreateAt:     time.Now(),
		UpdatedAt:    time.Now(),
	}
}
